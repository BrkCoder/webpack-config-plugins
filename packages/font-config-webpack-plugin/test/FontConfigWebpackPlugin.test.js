const path = require('path');
const rimraf = require('rimraf');
const webpack = require('webpack');
const FontConfigWebpackPlugin = require('../src/FontConfigWebpackPlugin');

// declare global test timeout to 10s for async tests
jest.setTimeout(20 * 1000);

const runWithinWebpackContext = (callback, settings = {}) => {
	webpack({
		mode: settings.mode || 'development',
		entry: path.resolve(__dirname, 'fixtures/entry.js'),
		target: 'node',
		node: {
			global: false,
		},
		output: {
			path: path.resolve(__dirname, 'fixtures/dist'),
			filename: 'bundle.js',
			libraryTarget: 'umd',
		},
		module: {
			rules: [
				{
					test: /\.css$/,
					use: ['style-loader', 'css-loader'],
				},
			],
		},
		resolve: {
			extensions: ['.js', '.css'],
		},
		plugins: [new FontConfigWebpackPlugin(settings.pluginOptions)],
	}).run((_, stats) => {
		callback({
			_,
			stats,
			options: stats.compilation.options,
			errors: stats.compilation.errors,
		});
	});
};

afterEach(resolve => {
	jest.resetModules();
	rimraf(path.join(__dirname, 'fixtures/dist'), resolve);
});

describe('FontConfigWebpackPlugin', () => {
	it('should be creatable without options', () => {
		new FontConfigWebpackPlugin();
	});

	describe('options', () => {
		it('options should match the last options snapshot (dev)', resolve => {
			runWithinWebpackContext(({ options }) => {
				expect(options).toMatchSnapshot();
				resolve();
			});
		});

		it('options should match the last custom options snapshot (dev)', resolve => {
			runWithinWebpackContext(
				({ options }) => {
					expect(options).toMatchSnapshot();
					resolve();
				},
				{
					pluginOptions: {
						name: '[name]-[hash].[ext]',
					},
				}
			);
		});

		// TODO: This test throws timeout error!
		it.skip(
			'options should match the last options snapshot (prod)',
			resolve => {
				runWithinWebpackContext(({ options }) => {
					expect(options).toMatchSnapshot();
					resolve();
				});
			},
			{ mode: 'production' }
		);
	});

	describe('runnable webpack context', () => {
		it('should finish without errors', () => {
			runWithinWebpackContext(({ errors }) => {
				expect(errors).toEqual([]);
			});
		});

		it('should set plugins correctly', resolve => {
			runWithinWebpackContext(({ options }) => {
				expect(options.plugins.length).toBe(1);
				expect(options.plugins[0] instanceof FontConfigWebpackPlugin).toBe(true);
				resolve();
			});
		});

		it('should set rules correctly', resolve => {
			runWithinWebpackContext(({ options }) => {
				const ruleToTest = options.module.rules[1];
				expect(options.module.rules.length - 1).toBe(1); // -1 is the rule defined in the test config
				expect(ruleToTest.test.test('webfont.eot')).toBe(true);
				expect(ruleToTest.test.test('webfont.ttf')).toBe(true);
				expect(ruleToTest.test.test('webfont.woff')).toBe(true);
				expect(ruleToTest.test.test('webfont.woff2')).toBe(true);
				expect(ruleToTest.test.test('webfont.svg')).toBe(true);
				resolve();
			});
		});
	});
});
