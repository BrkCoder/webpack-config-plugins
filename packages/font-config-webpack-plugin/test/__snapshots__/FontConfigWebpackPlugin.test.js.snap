// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontConfigWebpackPlugin options options should match the last custom options snapshot (dev) 1`] = `
Object {
  "cache": true,
  "context": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin",
  "devtool": "eval",
  "entry": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/test/fixtures/entry.js",
  "mode": "development",
  "module": Object {
    "defaultRules": Array [
      Object {
        "resolve": Object {},
        "type": "javascript/auto",
      },
      Object {
        "resolve": Object {
          "mainFields": Array [
            "main",
          ],
        },
        "test": /\\\\\\.mjs\\$/i,
        "type": "javascript/esm",
      },
      Object {
        "test": /\\\\\\.json\\$/i,
        "type": "json",
      },
      Object {
        "test": /\\\\\\.wasm\\$/i,
        "type": "webassembly/experimental",
      },
    ],
    "exprContextCritical": true,
    "exprContextRecursive": true,
    "exprContextRegExp": false,
    "exprContextRequest": ".",
    "rules": Array [
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
      Object {
        "test": /\\\\\\.\\(woff\\(2\\)\\?\\|ttf\\|eot\\|svg\\)\\(\\\\\\?v=\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\)\\?\\$/,
        "use": Array [
          Object {
            "loader": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/node_modules/cache-loader/dist/cjs.js",
            "options": Object {
              "cacheDirectory": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/node_modules/cache-loader/.cache-loader",
            },
          },
          Object {
            "loader": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "name": undefined,
              "outputPath": undefined,
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
    "strictThisContextOnImports": false,
    "unknownContextCritical": true,
    "unknownContextRecursive": true,
    "unknownContextRegExp": false,
    "unknownContextRequest": ".",
    "unsafeCache": true,
    "wrappedContextCritical": false,
    "wrappedContextRecursive": true,
    "wrappedContextRegExp": /\\.\\*/,
  },
  "node": Object {
    "Buffer": true,
    "__dirname": "mock",
    "__filename": "mock",
    "console": false,
    "global": false,
    "process": true,
    "setImmediate": true,
  },
  "optimization": Object {
    "checkWasmTypes": false,
    "concatenateModules": false,
    "flagIncludedChunks": false,
    "mangleWasmImports": false,
    "mergeDuplicateChunks": true,
    "minimize": false,
    "minimizer": Array [
      Object {
        "apply": [Function],
      },
    ],
    "namedChunks": true,
    "namedModules": true,
    "noEmitOnErrors": false,
    "nodeEnv": "development",
    "occurrenceOrder": false,
    "portableRecords": false,
    "providedExports": true,
    "removeAvailableModules": true,
    "removeEmptyChunks": true,
    "runtimeChunk": undefined,
    "sideEffects": false,
    "splitChunks": Object {
      "automaticNameDelimiter": "~",
      "cacheGroups": Object {
        "default": Object {
          "minChunks": 2,
          "priority": -20,
          "reuseExistingChunk": true,
        },
        "vendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "maxAsyncRequests": Infinity,
      "maxInitialRequests": Infinity,
      "minChunks": 1,
      "minSize": 10000,
      "name": true,
    },
    "usedExports": false,
  },
  "output": Object {
    "chunkCallbackName": "webpackChunk",
    "chunkFilename": "[id].bundle.js",
    "chunkLoadTimeout": 120000,
    "crossOriginLoading": false,
    "devtoolLineToLine": false,
    "devtoolNamespace": "",
    "filename": "bundle.js",
    "globalObject": "global",
    "hashDigest": "hex",
    "hashDigestLength": 20,
    "hashFunction": "md4",
    "hotUpdateChunkFilename": "[id].[hash].hot-update.js",
    "hotUpdateFunction": "webpackHotUpdate",
    "hotUpdateMainFilename": "[hash].hot-update.json",
    "jsonpFunction": "webpackJsonp",
    "jsonpScriptType": false,
    "library": "",
    "libraryTarget": "umd",
    "path": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/test/fixtures/dist",
    "pathinfo": true,
    "sourceMapFilename": "[file].map[query]",
    "strictModuleExceptionHandling": false,
    "webassemblyModuleFilename": "[modulehash].module.wasm",
  },
  "performance": false,
  "plugins": Array [
    FontConfigWebpackPlugin {
      "options": Object {
        "name": "[name]-[hash].[ext]",
      },
    },
  ],
  "resolve": Object {
    "aliasFields": Array [],
    "cacheWithContext": false,
    "extensions": Array [
      ".js",
      ".css",
      "woff",
      "woff2",
      "ttf",
      "eot",
      "svg",
    ],
    "mainFields": Array [
      "module",
      "main",
    ],
    "mainFiles": Array [
      "index",
    ],
    "modules": Array [
      "node_modules",
    ],
    "unsafeCache": true,
  },
  "resolveLoader": Object {
    "cacheWithContext": false,
    "extensions": Array [
      ".js",
      ".json",
    ],
    "mainFields": Array [
      "loader",
      "main",
    ],
    "mainFiles": Array [
      "index",
    ],
    "unsafeCache": true,
  },
  "target": "node",
}
`;

exports[`FontConfigWebpackPlugin options options should match the last options snapshot (dev) 1`] = `
Object {
  "cache": true,
  "context": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin",
  "devtool": "eval",
  "entry": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/test/fixtures/entry.js",
  "mode": "development",
  "module": Object {
    "defaultRules": Array [
      Object {
        "resolve": Object {},
        "type": "javascript/auto",
      },
      Object {
        "resolve": Object {
          "mainFields": Array [
            "main",
          ],
        },
        "test": /\\\\\\.mjs\\$/i,
        "type": "javascript/esm",
      },
      Object {
        "test": /\\\\\\.json\\$/i,
        "type": "json",
      },
      Object {
        "test": /\\\\\\.wasm\\$/i,
        "type": "webassembly/experimental",
      },
    ],
    "exprContextCritical": true,
    "exprContextRecursive": true,
    "exprContextRegExp": false,
    "exprContextRequest": ".",
    "rules": Array [
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
      Object {
        "test": /\\\\\\.\\(woff\\(2\\)\\?\\|ttf\\|eot\\|svg\\)\\(\\\\\\?v=\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\)\\?\\$/,
        "use": Array [
          Object {
            "loader": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/node_modules/cache-loader/dist/cjs.js",
            "options": Object {
              "cacheDirectory": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/node_modules/cache-loader/.cache-loader",
            },
          },
          Object {
            "loader": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "name": "[name].[ext]",
              "outputPath": "fonts/",
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
    "strictThisContextOnImports": false,
    "unknownContextCritical": true,
    "unknownContextRecursive": true,
    "unknownContextRegExp": false,
    "unknownContextRequest": ".",
    "unsafeCache": true,
    "wrappedContextCritical": false,
    "wrappedContextRecursive": true,
    "wrappedContextRegExp": /\\.\\*/,
  },
  "node": Object {
    "Buffer": true,
    "__dirname": "mock",
    "__filename": "mock",
    "console": false,
    "global": false,
    "process": true,
    "setImmediate": true,
  },
  "optimization": Object {
    "checkWasmTypes": false,
    "concatenateModules": false,
    "flagIncludedChunks": false,
    "mangleWasmImports": false,
    "mergeDuplicateChunks": true,
    "minimize": false,
    "minimizer": Array [
      Object {
        "apply": [Function],
      },
    ],
    "namedChunks": true,
    "namedModules": true,
    "noEmitOnErrors": false,
    "nodeEnv": "development",
    "occurrenceOrder": false,
    "portableRecords": false,
    "providedExports": true,
    "removeAvailableModules": true,
    "removeEmptyChunks": true,
    "runtimeChunk": undefined,
    "sideEffects": false,
    "splitChunks": Object {
      "automaticNameDelimiter": "~",
      "cacheGroups": Object {
        "default": Object {
          "minChunks": 2,
          "priority": -20,
          "reuseExistingChunk": true,
        },
        "vendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "maxAsyncRequests": Infinity,
      "maxInitialRequests": Infinity,
      "minChunks": 1,
      "minSize": 10000,
      "name": true,
    },
    "usedExports": false,
  },
  "output": Object {
    "chunkCallbackName": "webpackChunk",
    "chunkFilename": "[id].bundle.js",
    "chunkLoadTimeout": 120000,
    "crossOriginLoading": false,
    "devtoolLineToLine": false,
    "devtoolNamespace": "",
    "filename": "bundle.js",
    "globalObject": "global",
    "hashDigest": "hex",
    "hashDigestLength": 20,
    "hashFunction": "md4",
    "hotUpdateChunkFilename": "[id].[hash].hot-update.js",
    "hotUpdateFunction": "webpackHotUpdate",
    "hotUpdateMainFilename": "[hash].hot-update.json",
    "jsonpFunction": "webpackJsonp",
    "jsonpScriptType": false,
    "library": "",
    "libraryTarget": "umd",
    "path": "/Users/janbiasi/github/namics-opensource/webpack-config-plugins/packages/font-config-webpack-plugin/test/fixtures/dist",
    "pathinfo": true,
    "sourceMapFilename": "[file].map[query]",
    "strictModuleExceptionHandling": false,
    "webassemblyModuleFilename": "[modulehash].module.wasm",
  },
  "performance": false,
  "plugins": Array [
    FontConfigWebpackPlugin {
      "options": Object {
        "format": "[name].[ext]",
        "output": "fonts/",
      },
    },
  ],
  "resolve": Object {
    "aliasFields": Array [],
    "cacheWithContext": false,
    "extensions": Array [
      ".js",
      ".css",
      "woff",
      "woff2",
      "ttf",
      "eot",
      "svg",
    ],
    "mainFields": Array [
      "module",
      "main",
    ],
    "mainFiles": Array [
      "index",
    ],
    "modules": Array [
      "node_modules",
    ],
    "unsafeCache": true,
  },
  "resolveLoader": Object {
    "cacheWithContext": false,
    "extensions": Array [
      ".js",
      ".json",
    ],
    "mainFields": Array [
      "loader",
      "main",
    ],
    "mainFiles": Array [
      "index",
    ],
    "unsafeCache": true,
  },
  "target": "node",
}
`;
